src/backend/distributed/executor/local_executor.c

```diff
@@ -229,6 +235,131 @@ ExtractParametersForLocalExecution(ParamListInfo paramListInfo, Oid **parameterT
 }
 
 
+/*
+ * ExecuteLocalUtilityTaskList executes a list of tasks locally. This function
+ * also logs local execution notice for each task and sets
+ * TransactionAccessedLocalPlacement to true for next set of possible queries
+ * & commands within the current transaction block. See the comment in function.
+ */
+void
+ExecuteLocalUtilityTaskList(List *localTaskList)
+{
+	Task *localTask = NULL;
+
+	foreach_ptr(localTask, localTaskList)
+	{
+		const char *localTaskQueryCommand = TaskQueryString(localTask);
+
+		if (localTask->anchorShardId != INVALID_SHARD_ID)
+		{
+			/*
+			 * If anchor shard id is valid, then we really access a local
+			 * placement. We should register it to force the local execution
+			 * of the following commands withing the current transcation.
+			 * It can be a coordinated transaction or a transaction that is
+			 * initiated explicitly.
+			 */
+			TransactionAccessedLocalPlacement = true;
+		}
+
+		LogLocalCommand(localTask);
+
+		/* execute the local utility task finally */
+		LocallyExecuteUtilityTask(localTaskQueryCommand);
+	}
+}
+
+
+/*
+ * LocallyExecuteUtilityTask executes the given local task query. Given
+ * query would mostly be a utility command as the name of the function
+ * implies. However, some utility commands can trigger udf calls to finish
+ * their execution. See the comment in function.
+ */
+static void
+LocallyExecuteUtilityTask(const char *localTaskQueryCommand)
+{
+	RawStmt *localTaskRawStmt = (RawStmt *) ParseTreeRawStmt(localTaskQueryCommand);
+
+	Node *localTaskRawParseTree = localTaskRawStmt->stmt;
+
+	/*
+	 * Actually, the query passed to this function would mostly be a
+	 * utility command to be executed locally. However, some utility
+	 * commands do trigger udf calls (e.g worker_apply_shard_ddl_command)
+	 * to execute commands in a generic way. But as we support local
+	 * execution of utility tasks, we should also process those udf
+	 * calls locally as well. In that case, we simply execute the query
+	 * implying the udf call in below conditional block.
+	 */
+
+	if (IsA(localTaskRawParseTree, SelectStmt))
+	{
+		SelectStmt *selectStmt = (SelectStmt *) localTaskRawParseTree;
```

Comment:
I think it'd make it easier to read the code if we can move the code of the `if` block into a function, because most of the work we do is to check for `Assert(false);`, so nothing very interesting in this function

Reply:
Or moving the part that I perform `foundNonUdfSelectTarget` check into another function like `ErrorIfNonUdf...` ?

Reply:
moving the assertion and the checks in this block into the `LocallyExecuteUdfTaskQuery` makes sense to me. Does it sound good ?


`

Reply:
moved the the part validating the `SELECT udf` call into another function